File: audiobridge/android/app/src/main/java/app/audiobridge/net/DiscoveryResponder.kt
package app.audiobridge.net

import org.json.JSONObject
import java.net.DatagramPacket
import java.net.DatagramSocket
import kotlin.concurrent.thread

/**
 * 디스커버리 응답기 (PROTOCOL.md §1) — 앱이 켜져 있는 동안 이 폰을 찾을 수 있게 한다.
 * 정확한 "ABDISC?" 질의에만, 질의를 보낸 주소로만 응답한다.
 * 이름은 매 응답 시점에 [nameProvider]에서 읽으므로 이름 변경이 즉시 반영된다.
 */
class DiscoveryResponder(private val nameProvider: () -> String, private val ctlPort: Int) {
    private var socket: DatagramSocket? = null
    @Volatile private var running = false

    fun start(): Boolean {
        val s = try {
            DatagramSocket(Protocol.DISCOVERY_PORT)
        } catch (_: Exception) {
            return false
        }
        socket = s
        running = true
        thread(name = "ab-disc-resp") {
            val buf = ByteArray(64)
            while (running) {
                try {
                    val pkt = DatagramPacket(buf, buf.size)
                    s.receive(pkt)
                    if (pkt.length == 7 && String(pkt.data, 0, 7, Charsets.US_ASCII) == "ABDISC?") {
                        val reply = ("ABDISC!" + JSONObject()
                            .put("name", nameProvider())
                            .put("version", Protocol.VERSION)
                            .put("ctlPort", ctlPort)
                            .toString()).toByteArray(Charsets.UTF_8)
                        s.send(DatagramPacket(reply, reply.size, pkt.address, pkt.port))
                    }
                } catch (_: Exception) {
                    if (!running) return@thread
                }
            }
        }
        return true
    }

    fun stop() {
        running = false
        runCatching { socket?.close() }
        socket = null
    }
}

File: audiobridge/windows/DiscoveryResponder.cs
using System.Net;
using System.Net.Sockets;
using System.Text;
using System.Text.Json;

namespace AudioBridge.Win;

/// <summary>UDP 48551에서 "ABDISC?" 질의에 유니캐스트로 응답한다 (PROTOCOL.md §1).</summary>
public sealed class DiscoveryResponder : IDisposable
{
    private readonly int _ctlPort;
    private UdpClient? _udp;
    private volatile bool _running;

    public DiscoveryResponder(int ctlPort)
    {
        _ctlPort = ctlPort;
    }

    public bool Start()
    {
        try
        {
            _udp = new UdpClient(Protocol.DiscoveryPort);
        }
        catch (SocketException)
        {
            Console.WriteLine($"[검색] 포트 {Protocol.DiscoveryPort} 사용 중 — 자동 검색이 꺼져요 (폰에서 IP 직접 입력은 가능)");
            return false;
        }
        _running = true;
        var t = new Thread(Loop) { IsBackground = true, Name = "ab-discovery" };
        t.Start();
        Console.WriteLine("[검색] 폰의 검색에 응답할 준비 완료");
        return true;
    }

    private void Loop()
    {
        while (_running)
        {
            try
            {
                var ep = new IPEndPoint(IPAddress.Any, 0);
                byte[] data = _udp!.Receive(ref ep);
                // 정확한 질의에만, 질의를 보낸 주소로만 응답 (증폭/스캔 방지)
                if (data.Length == 7 && Encoding.ASCII.GetString(data) == "ABDISC?")
                {
                    var reply = Encoding.UTF8.GetBytes(
                        "ABDISC!" + JsonSerializer.Serialize(new { name = Config.Name, version = Protocol.Version, ctlPort = _ctlPort }));
                    _udp.Send(reply, reply.Length, ep);
                }
            }
            catch
            {
                if (!_running) return;
            }
        }
    }

    public void Dispose()
    {
        _running = false;
        _udp?.Dispose();
    }
}
